Feat/switch groth16 key

lucasmenendez opened the pull request:
# Description

This PR implements `SwitchVerificationKey` for groth16 recursion. By including two verification keys for similar circuits, it allows recursive verification of multiple proofs for both circuits, seamlessly switching between their corresponding verification keys as needed.


## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How has this been tested?

There are three similar tests that verify that two proofs from two different circuits are in the third circuit. Each test uses different curves for the inner and outer circuits:

- [x] Test [`TestMultipleBN254InBN254`](https://github.com/lucasmenendez/gnark/blob/becbd0e43531ea2cd0c4b23d0da60f4852db5398/std/recursion/groth16/verifier_test.go#L546)
- [x] Test [`TestMultipleBLS12InBW6`](https://github.com/lucasmenendez/gnark/blob/becbd0e43531ea2cd0c4b23d0da60f4852db5398/std/recursion/groth16/verifier_test.go#L591)
- [x] Test [`TestMultipleBW6InBN254`](https://github.com/lucasmenendez/gnark/blob/becbd0e43531ea2cd0c4b23d0da60f4852db5398/std/recursion/groth16/verifier_test.go#L637)

# Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I did not modify files generated from templates
- [x] `golangci-lint` does not output errors locally
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

ivokub commented:
I'm a bit behind on reviews and will get to it soon

ivokub commented:
> I'm a bit behind on reviews and will get to it soon

I rebased on top of master and fixed conflicting changes. I'll still have a look at the Groth16 test and add tests for MuxG2/MuxGt. And I'd like to write documentation example. But in general looks good.

lucasmenendez commented:
Super nice!!

> > I'm a bit behind on reviews and will get to it soon
> 
> I rebased on top of master and fixed conflicting changes. I'll still have a look at the Groth16 test and add tests for MuxG2/MuxGt. And I'd like to write documentation example. But in general looks good.

ivokub commented:
Thanks! It is super useful. It also revealed a bug we had before.
